cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/reachability/ReachableBlock.qll

```diff
@@ -0,0 +1,52 @@
+private import ReachableBlockInternal
+private import semmle.code.cpp.ir.internal.IntegerConstant
+private import IR
+private import ConstantAnalysis
+
+predicate isInfeasibleEdge(IRBlock block, EdgeKind kind) {
+  exists(ConditionalBranchInstruction instr, int conditionValue |
+    instr = block.getLastInstruction() and
+    conditionValue = getValue(getConstantValue(instr.getCondition())) and
+    if conditionValue = 0 then
+      kind instanceof TrueEdge
+    else
+      kind instanceof FalseEdge
+  )
+}
+
+IRBlock getAFeasiblePredecessor(IRBlock successor) {
+  exists(EdgeKind kind |
+    result.getSuccessor(kind) = successor and
+    not isInfeasibleEdge(result, kind)
+  )
+}
+
+predicate isBlockReachable(IRBlock block) {
+  getAFeasiblePredecessor*(block) = block.getFunctionIR().getEntryBlock()
```

Comment:
The optimiser is often unlucky on joins like this one. Could `block.getFunctionIR()` be replaced with `any(FunctionIR f)`?

Reply:
Done